dask/array/core.py

```diff
@@ -2566,29 +2609,61 @@ def concatenate(seq, axis=0, allow_unknown_chunksizes=False):
     return Array(dsk2, name, chunks, dtype=dt)
 
 
-def insert_to_ooc(out, arr, lock=True, region=None):
+def insert_to_ooc(arr, out, lock=True, region=None, keep=False):
     if lock is True:
         lock = Lock()
 
-    def store(out, x, index, lock, region):
+    def store(x, out, index, lock, region, keep):
+        result = None
+        if keep:
+            result = out
+
+        subindex = index
+        if region is not None:
+            subindex = fuse_slice(region, index)
+
         if lock:
             lock.acquire()
         try:
-            if region is None:
-                out[index] = np.asanyarray(x)
-            else:
-                out[fuse_slice(region, index)] = np.asanyarray(x)
+            out[subindex] = np.asanyarray(x)
         finally:
             if lock:
                 lock.release()
 
-        return None
+        return result
 
     slices = slices_from_chunks(arr.chunks)
 
-    name = 'store-%s' % arr.name
-    dsk = {(name,) + t[1:]: (store, out, t, slc, lock, region)
-           for t, slc in zip(core.flatten(arr.__dask_keys__()), slices)}
+    dsk = dict()
+    for t, slc in zip(core.flatten(arr.__dask_keys__()), slices):
+        store_key = ('store-%s' % arr.name,) + t[1:]
+        dsk[store_key] = (store, t, out, slc, lock, region, keep)
+
+    return dsk
+
+
+def retrieve_from_ooc(out_dsk):
+    def load(x, index, lock, region):
+        result = None
+
+        subindex = index
+        if region is not None:
+            subindex = fuse_slice(region, index)
+
+        if lock:
+            lock.acquire()
+        try:
+            result = x[subindex]
+        finally:
+            if lock:
+                lock.release()
+
+        return result
+
+    dsk = dict()
+    for store_key in out_dsk:
+        load_key = ('load-%s' % store_key[0].lstrip("store-"),) + store_key[1:]
+        dsk[load_key] = (load, store_key,) + out_dsk[store_key][3:-1]
 
     return dsk
```

Comment:
Why is this necessary on top of `from_array`?  They seem very similar

Reply:
😕 I don't think we are calling `from_array` or are you meaning that you feel the behavior is similar to `from_array`? If the latter, more on this in the comment to follow.

Reply:
This logic seemed similar to `from_array`.  I was curious if there was a reason why we can't reuse the existing functionality.  Mostly I'm concerned with the amount of code being written into dask.array.core to support this feature.  If it isn't easy to accomplish what we want to do with existing tools then is there something we can do to improve our current tools?